ocf_data_sampler/torch_datasets/datasets/pvnet_uk.py

```
@@ -117,13 +117,48 @@ def __init__(
         self.means_dict = means_dict
         self.stds_dict = stds_dict
 
-    def process_and_combine_datasets(
+
+    def normalize_datasets(self, dataset_dict: dict) -> dict:
+        """Normalise the datasets.
+
+        Args:
+            dataset_dict: Dictionary of xarray datasets
+        """
+        if "nwp" in dataset_dict:
+
+            for nwp_key, da_nwp in dataset_dict["nwp"].items():
+
+                # Standardise in-place
+                channel_means = self.means_dict["nwp"][nwp_key]
+                channel_stds = self.stds_dict["nwp"][nwp_key]
+
+                da_nwp.values = (da_nwp.values - channel_means) / channel_stds
```

Is it correct to say that the in-place normalisation here means that xarray's broadcasting and dimension alignment checks won't happen so we would have to be sure that everything is in the correct order?

If so I couldn't see a test around `config_normalization_values_to_dicts` which is used above and seems to be the function which would define that order, should we add a test there or is there a test somewhere else that checks the ordering?

---

Since `channel_means` and `channel_stds` are just arrays, we were already skipping xarray's broadcasting and dimension alignment. I didn't quite look into everything that this skips past inside xarray, but I think in the old version there is internal copying happening inside the xarray __sub__ and __mult__ methods. Now we're just relying on the fast that `channel_means/stds` and `da_nwp.values` which are numpy arrays have compatible shapes for numpy broadcasting.

So no we don't have tests around `config_normalization_values_to_dicts` and the dimension order. Right now `channel_means/stds` [have shapes](https://github.com/openclimatefix/ocf-data-sampler/blob/main/ocf_data_sampler/torch_datasets/utils/config_normalization_values_to_dicts.py#L42-L43) [1, channels, 1, 1]. The sat data has [shape](https://github.com/openclimatefix/ocf-data-sampler/blob/main/ocf_data_sampler/load/satellite.py#L37) [time, channels, x,  y], and after time slicing the NWP has the same dimension order too. 

So the alignment in the shapes is there in that they need to be compatible arrays for numpy broadcasting. But yes this produce make a silent error if we transposed the dimensions by accident, and if the norm values had the same number of channels as another dimension of the sat/NWP. But it will fail loudly if the number of channels doesn't match. 

I think the risk is probably minimal, but I'm open to the idea that there could be a test. I'm just not sure what we'd be looking for in a test. Any thoughts?

---

Thanks for explaining, hmmm maybe we could throw in a check rather than a test, something like this 
```
if da_nwp.dims[1] != "channel":
    raise ValueError(f"Expected second dimension to be 'channel', but got '{da_nwp.dims[1]}'") 
```
and a comment in the `config_normalization_values_to_dicts` function that the order is there to match how we transpose the dimensions 

I agree the chance is small but it feels strange not add in some check for what could lead to a silent error, but I don't feel super strong on this so if you think this is overkill then we can skip it